dom/events/event-global.html

```diff
@@ -0,0 +1,92 @@
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<div id=log></div>
+<script>
+test(t => {
+  assert_own_property(window, "event");
+  assert_equals(window.event, undefined);
+}, "event exists on window, which is initially set to undefined");
+
+async_test(t => {
+  let target = document.createElement("div");
+  assert_equals(window.event, undefined, "undefined before dispatch");
+
+  target.addEventListener("click", t.step_func_done(e => {
+    assert_not_equals(window.event, undefined, "not undefined during dispatch");
+  }));
+
+  target.dispatchEvent(new Event("click"));
+  assert_equals(window.event, undefined, "undefined after dispatch");
+}, "window.event is only defined during dispatch");
+
+async_test(t => {
+  let parent = document.createElement("div");
+  let root = parent.attachShadow({mode: "closed"});
+  let span = document.createElement("span");
+  root.appendChild(span);
+
+  span.addEventListener("test", t.step_func(e => {
+    assert_equals(window.event, undefined);
+    assert_not_equals(window.event, e);
+  }));
+
+  parent.addEventListener("test", t.step_func_done(e => {
+    assert_equals(window.event, e);
+    assert_not_equals(window.event, undefined);
+  }));
+
+  parent.dispatchEvent(new Event("test", {composed: true}));
+}, "window.event is undefined if the target is in a shadow tree, when event is dispatched outside shadow tree");
+
+async_test(t => {
+  let parent = document.createElement("div");
+  let root = parent.attachShadow({mode: "closed"});
+  let span = document.createElement("span");
+  root.appendChild(span);
+  let shadowNode = root.firstElementChild;
+
+  shadowNode.addEventListener("test", t.step_func((e) => {
+    assert_not_equals(window.event, e);
+    assert_equals(window.event, undefined);
+  }));
+
+  parent.addEventListener("test", t.step_func_done(e => {
+    assert_equals(window.event, e);
+    assert_not_equals(window.event, undefined);
+  }));
+
+  shadowNode.dispatchEvent(new Event("test", {composed: true, bubbles: true}));
+}, "window.event is undefined if the target is in a shadow tree, when event is dispatched inside shadow tree");
+
+async_test(t => {
+  let target1 = document.createElement("div");
+  let target2 = document.createElement("div");
+
+  target2.addEventListener("dude", t.step_func(() => {
+    assert_not_equals(window.event.type, "cool");
+    assert_equals(window.event.type, "dude");
+  }));
+
+  target1.addEventListener("cool", t.step_func_done(() => {
+    assert_equals(window.event.type, "cool", "got expected event from global event during dispatch");
+    target2.dispatchEvent(new Event("dude"));
+    assert_not_equals(window.event.type, "dude", "got expected event from global event after handling a different event handler callback");
+    assert_equals(window.event.type, "cool", "got expected event from global event after handling a different event handler callback");
```

Comment:
Doesn't the spec PR say that this should be undefined? If window.event needs to be able to change to different events and back, maybe it needs to be a stack (internally)?

Reply:
Let me re-publish an email exchange I had with @foolip:

> On Jan 26, 2018, at 2:51 AM, Philip Jägenstedt <...> wrote:
> Here's the code in Chromium:
https://cs.chromium.org/chromium/src/third_party/WebKit/Source/bindings/core/v8/custom/V8WindowCustom.cpp
> Looks like this is stored as private property on each window instance, in other words not a stack.
> 
> In https://cs.chromium.org/chromium/src/third_party/WebKit/Source/bindings/core/v8/V8AbstractEventListener.cpp, the old value is saved and then restored after the event listener was invoked.
>
> This looks roughly like step 2.7 of https://dom.spec.whatwg.org/#concept-event-listener-inner-invoke. If there's any way to observe the difference between doing this around each individual invoke, or for the whoe dispatch, that matters. (I can't think of a way though.)

On Thu, Jan 25, 2018 at 1:07 AM Mike Taylor <...> wrote:
>> So here’s my understanding:
>>
>> During event dispatch:
>> * set window.event to the current event (unless we’re in shadow doc land, we leave it as undefined).
>>
>> What I’m not sure how to define is the scenario where it’s possible to dispatch an event from a listener’s callback — so we need to change window.event to the new current event during that dispatch, and then restore it back to the old current event (when we return to the callback from inner invoke).
>>
>> So… should we define this as a stack? Or am I overthinking this?

Reply:
OK, re-reading that and looking at https://whatpr.org/dom/407.html#ref-for-concept-event-listener-inner-invoke:

If I set *current event* to `event` in between Steps 1 and 2 of inner invoke, and set it to undefined after Step 2, this should work, I think.

Thinking out loud:

When `cool` is dispatched:
*  do the `cool` inner invoke and set *current event* to `cool`

During the `cool` listener's callback, `dude` is dispatched:
*  do the `dude` inner invoke and set *current event* to `dude`

Once that is done, resume step 7 of the `cool` inner invoke and its *current event* (`cool`) is in scope again, right?

Reply:
Yeah... except instead of setting to undefined afterwards, restore to the saved value (might be undefined or an event).